File: examples/sfm_global.py
import cv2
import numpy as np
import open3d as o3d
from scipy.optimize import least_squares
from scipy.spatial.transform import Rotation
from scipy.sparse import lil_matrix
from copy import deepcopy
import time
class ProjectError():
    def __init__(self,n_cameras, n_points):
        self.n_cameras = n_cameras
        self.n_points = n_points

    def sparsity_jacobian_inc(self, camera_indices, point_indices):
        
        m = camera_indices.size * 2 # m = 800 x 2
        n = 6 + self.n_points * 3 # n = 6 * cameras = 5 + 3 * points = 160 --> 510
        A = lil_matrix((m, n), dtype=int)

        i = np.arange(camera_indices.size)
        for s in range(6):
            A[2 * i, camera_indices * 6 + s] = 1
            A[2 * i + 1, camera_indices * 6 + s] = 1

        for s in range(3):
            A[2 * i, 6 + point_indices * 3 + s] = 1
            A[2 * i + 1, 6 + point_indices * 3 + s] = 1

        return A

    def sparsity_jacobian_global(self, camera_indices, point_indices):
        # camera, point indices = params to 2d points
        m = camera_indices.size * 2 # m = 800 x 2
        n = 9 * self.n_cameras + self.n_points * 3 # n = 6 * cameras = 5 + 3 * points = 160 --> 510
        A = lil_matrix((m, n), dtype=int)

        i = np.arange(camera_indices.size)
        for s in range(9):
            A[2 * i, camera_indices * 9 + s] = 1
            A[2 * i + 1, camera_indices * 9 + s] = 1

        for s in range(3):
            A[2 * i, 9  + point_indices * 3 + s] = 1
            A[2 * i + 1, 9 + point_indices * 3 + s] = 1

        return A

    def project2(self, cam_params, points):
        points_proj = []
        for i in range(len(cam_params)):
            # Init params
            f, cx, cy = cam_params[i, 0], cam_params[i, 1], cam_params[i, 2]
            rvec, tvec = cam_params[i, 3:6], cam_params[i, 6:9]
            point_3d = points[i]
            K = np.array([[f, 0, cx], [0, f, cy], [0, 0, 1]], dtype=np.float32)

            # Do Projection
            rot = Rotation.from_rotvec(rvec)
            point_proj = rot.apply(point_3d)
            point_proj += tvec
            point_proj = point_proj @ K.T
            point_proj[0] /= point_proj[2]
            point_proj[1] /= point_proj[2]
            points_proj.append(point_proj[:2])
        return points_proj


    @staticmethod
    def cost_func(params, points_2d, sfm, cam_indices, point_indices,cls):
        """
        params = (camera_params + points_3d).ravel(). 
        n_cameras = cameras number, which is also index of scene
        cls is ba itself
        camera params = f, cx, cy, rx, ry, rz, tx, ty, tz
        """
        cam_params = params[: cls.n_cameras * 9].reshape((cls.n_cameras, 9))
        points_3d = params[cls.n_cameras * 9:].reshape((cls.n_points, 3))

        proj_points_3d = cls.project2(cam_params[cam_indices], points_3d[point_indices])
        result = points_2d - proj_points_3d
        return result.ravel()

class SFM():
    """3D point idx Gen data structure"""
    def __init__(self):
        self._sfm_dict = {}

    def gen_dict(self, cam_idx, point_2d_idx):
        """
        Gen data set. 
        (Cam idx, 2D Pts idx) = 3D Pts idx
        3D Pts Idx = (Cam idx, 2D Pts idx) --> find 하기 어려울 수 있음.
        """
        if self._sfm_dict.get((cam_idx, point_2d_idx)) is None:
            self._sfm_dict[(cam_idx, point_2d_idx)] = None

        return self._sfm_dict[(cam_idx, point_2d_idx)]

    def pop(self, key):
        self._sfm_dict.pop(key)

    def gen_indices(self, img_keypoints):
        cam_indices = []
        point_indices = []
        points_2d = []
        for (cam_idx, m_idx), p3_idx in self._sfm_dict.items():
            points_2d.append(img_keypoints[cam_idx][m_idx].pt)
            cam_indices.append(cam_idx)
            point_indices.append(p3_idx)
        return np.array(cam_indices), np.array(point_indices), np.array(points_2d)

def main():
    img_path = "../bin/data/relief/%02d.jpg"
    img_resize = 0.25
    f_init, cx_init, cy_init, Z_init, Z_limit = 500, -1, -1, 2, 100
    ba_loss_width = 9
    min_inlier_num, ba_num_iter = 200, 200
    SHOW_MATCH = False

    # Load images and extract features
    # USE BRISK feature detection
    img_keypoints = []
    img_descriptors = []
    img_set = []

    detector = cv2.BRISK_create()

    cam = cv2.VideoCapture(img_path)
    h, w = 0, 0
    while True:
        _, img = cam.read()
        if img is None: break
        img = cv2.resize(img, dsize=(0, 0), fx=img_resize, fy=img_resize)
        img_keypoint, img_descriptor = detector.detectAndCompute(img, None)
        img_keypoints.append(img_keypoint)
        img_descriptors.append(img_descriptor)
        img_set.append(img)
        h, w, _ = img.shape
    cam.release()
    if cx_init < 0 or cy_init < 0:
        cx_init = w / 2
        cy_init = h / 2

    img_keypoints = np.array(img_keypoints, dtype=object)
    img_descriptors = np.array(img_descriptors, dtype=object)
    img_set = np.array(img_set)
    
    # Match features and find good matches
    fmatcher = cv2.DescriptorMatcher_create("BruteForce-Hamming")
    match_pair, match_inlier = [], []
    for i in range(len(img_set)):
        for j in range(i+1, len(img_set)):
            src, dst, inlier = [], [], []
            match = fmatcher.match(img_descriptors[i], img_descriptors[j])
            match = np.array(match)
            for m in match:
                src.append(img_keypoints[i][m.queryIdx].pt)
                dst.append(img_keypoints[j][m.trainIdx].pt)
            
            src = np.array(src, dtype=np.float32)
            dst = np.array(dst, dtype=np.float32)
            
            F, inlier_mask = cv2.findFundamentalMat(src, dst, cv2.RANSAC)
            for k in range(len(inlier_mask)):
                if inlier_mask[k]: 
                    inlier.append(match[k]) # 매칭된 index를 넣음.
            inlier = np.array(inlier)

            print(f"3DV Tutorial: Image {i} - {j} are matched ({inlier.size} / {match.size}).\n")

            # Determin whether the image pair is good or not
            if inlier.size < min_inlier_num: continue
            print(f"3DV Tutorial: Image {i} - {j} are selected.\n")
            match_pair.append((i, j))
            match_inlier.append(inlier)

            if SHOW_MATCH:
                match_image = cv2.drawMatches(img_set[i], img_keypoints[i], img_set[j], img_keypoints[j], match, (0, 255, 0), (255, 0, 0), matchesMask=inlier_mask)
                cv2.imshow("3DV Tutorial: Structure-from-Motion", match_image)
                cv2.waitKey()
    # if match_pair.size < 1: return 
    
    # Find 0 - 1 - 2 Covisibility Matched points
    points_3d = []
    points_rgb = []
    xs_visited = SFM()
    for i in range(len(match_pair)):
        for j in range(len(match_inlier[i])):
            cam1_idx, cam2_idx = match_pair[i][0], match_pair[i][1] 
            pts1_2d_idx, pts2_2d_idx = match_inlier[i][j].queryIdx, match_inlier[i][j].trainIdx

            value1 = xs_visited.gen_dict(cam_idx=cam1_idx, point_2d_idx=pts1_2d_idx)
            value2 = xs_visited.gen_dict(cam_idx=cam2_idx, point_2d_idx=pts2_2d_idx)

            # 둘 다 값이 있으면
            if value1 != None and value2 != None:
                # 근데 서로 이상한걸 가리키면
                if value1 != value2:
                    xs_visited.pop((cam1_idx, pts1_2d_idx))
                    xs_visited.pop((cam2_idx, pts2_2d_idx))
                
                continue

            X_idx = 0
            if value1 != None:
                X_idx = value1
            elif value2 != None:
                X_idx = value2
            else:
                X_idx = len(points_3d)
                points_3d.append(np.array([0, 0, Z_init]))
                rgb_p = img_keypoints[cam1_idx][pts1_2d_idx].pt
                points_rgb.append(img_set[cam1_idx][int(rgb_p[1]), int(rgb_p[0])])
            if value1 == None:
                xs_visited._sfm_dict[(cam1_idx, pts1_2d_idx)] = X_idx
            if value2 == None:
                xs_visited._sfm_dict[(cam2_idx, pts2_2d_idx)] = X_idx

    print(f"3DV Tutorial: # of 3D points: {len(points_3d)}")
    
    ### Init Parameters ###
    points_3d = np.array(points_3d, dtype=np.float32)

    n_point_3d = len(points_3d)
    n_cameras = len(img_set)
    init_param = np.array([f_init, cx_init, cy_init, 0, 0, 0, 0, 0, 0], dtype=np.float32)
    cam_params = np.full( (len(img_set), len(init_param)), init_param, dtype=np.float32)
    params = np.hstack((cam_params.ravel(), points_3d.ravel()))

    BA = ProjectError(n_cameras=n_cameras, n_points=n_point_3d)

    # Ravel 2d points
    cam_indices, point_indices, points_2d = xs_visited.gen_indices(img_keypoints=img_keypoints)
    J = BA.sparsity_jacobian_global(camera_indices=cam_indices, point_indices=point_indices)

    start = time.time()
    opt = least_squares(BA.cost_func, params, jac_sparsity=J, method='trf', ftol=1e-4, verbose = 2, args=(points_2d, xs_visited, cam_indices, point_indices, BA))
    total = time.time() - start
    print(f"Total time is {total:.3f}")

    # Mark errorneous points to reject them
    opt_cam_params = opt.x[:n_cameras * 9].reshape(n_cameras, 9)
    opt_points_3d = opt.x[n_cameras * 9:].reshape(n_point_3d, 3)

    for i in range(len(opt_cam_params)):
        msg = f"3DV Tutorial: Camera {i+1}'s (f, cx, cy) = {opt_cam_params[i][0]:.1f} {opt_cam_params[i][1]:.1f} {opt_cam_params[i][2]:.1f}"
        print(msg)
    
    # Store the 3D points to an XYZ file
    points_3d_name = "sfm_global(point).xyz"
    with open(points_3d_name, "wt") as f:
        for i in range(n_point_3d):
            data = f"{opt_points_3d[i,0]:.3f} {opt_points_3d[i,1]:.3f} {opt_points_3d[i,2]:.3f}\n"
            f.write(data)

    points_rgb_name = "sfm_global(rgb).xyz"
    with open(points_rgb_name, "wt") as f:
        for i in range(n_point_3d):
            data = f"{points_rgb[i][0]:.3f} {points_rgb[i][1]:.3f} {points_rgb[i][2]:.3f}\n"
            f.write(data)

    camera_file = "sfm_global(camera).xyz"
    with open(camera_file, 'wt') as f:
        for i in range(n_cameras):
            data = f"{opt_cam_params[i, 0]:.3f} {opt_cam_params[i, 1]:.3f} {opt_cam_params[i, 2]:.3f} {opt_cam_params[i, 3]:.3f} {opt_cam_params[i, 4]:.3f} {opt_cam_params[i, 5]:.3f}\n"
            f.write(data)
    
    print("END!!!")
if __name__ == "__main__":
    main()

File: examples/sfm_global.cpp
#include "sfm.hpp"

std::vector<bool> maskNoisyPoints(std::vector<cv::Point3d>& Xs, const std::vector<std::vector<cv::KeyPoint>>& xs, const std::vector<SFM::Vec9d>& views, const SFM::VisibilityGraph& visibility, double reproj_error2)
{
    std::vector<bool> is_noisy(Xs.size(), false);
    if (reproj_error2 > 0)
    {
        for (auto visible = visibility.begin(); visible != visibility.end(); visible++)
        {
            cv::Point3d& X = Xs[visible->second];
            if (X.z < 0) continue;
            int img_idx = SFM::getCamIdx(visible->first), pt_idx = SFM::getObsIdx(visible->first);
            const cv::Point2d& x = xs[img_idx][pt_idx].pt;
            const SFM::Vec9d& view = views[img_idx];

            // Project the given 'X'
            cv::Vec3d rvec(view[0], view[1], view[2]);
            cv::Matx33d R;
            cv::Rodrigues(rvec, R);
            cv::Point3d X_p = R * X + cv::Point3d(view[3], view[4], view[5]);
            const double &f = view[6], &cx = view[7], &cy = view[8];
            cv::Point2d x_p(f * X_p.x / X_p.z + cx, f * X_p.y / X_p.z + cy);

            // Calculate distance between 'x' and 'x_p'
            cv::Point2d d = x - x_p;
            if (d.x * d.x + d.y * d.y > reproj_error2) is_noisy[visible->second] = true;
        }
    }
    return is_noisy;
}

int main()
{
    const char* input = "../data/relief/%02d.jpg";
    double img_resize = 0.25, f_init = 500, cx_init = -1, cy_init = -1, Z_init = 2, Z_limit = 100, ba_loss_width = 9; // Negative 'loss_width' makes BA not to use a loss function.
    int min_inlier_num = 200, ba_num_iter = 200; // Negative 'ba_num_iter' uses the default value for BA minimization
    bool show_match = false;

    // Load images and extract features
    cv::VideoCapture video;
    if (!video.open(input)) return -1;
    cv::Ptr<cv::FeatureDetector> fdetector = cv::BRISK::create();
    std::vector<std::vector<cv::KeyPoint>> img_keypoint;
    std::vector<cv::Mat> img_set, img_descriptor;
    while (true)
    {
        cv::Mat image;
        video >> image;
        if (image.empty()) break;
        if (img_resize != 1) cv::resize(image, image, cv::Size(), img_resize, img_resize);

        std::vector<cv::KeyPoint> keypoint;
        cv::Mat descriptor;
        fdetector->detectAndCompute(image, cv::Mat(), keypoint, descriptor);
        img_set.push_back(image);
        img_keypoint.push_back(keypoint);
        img_descriptor.push_back(descriptor.clone());
    }
    if (img_set.size() < 2) return -1;
    if (cx_init < 0) cx_init = img_set.front().cols / 2;
    if (cy_init < 0) cy_init = img_set.front().rows / 2;

    // Match features and find good matches
    cv::Ptr<cv::DescriptorMatcher> fmatcher = cv::DescriptorMatcher::create("BruteForce-Hamming");
    std::vector<std::pair<uint, uint>> match_pair;        // Good matches (image pairs)
    std::vector<std::vector<cv::DMatch>> match_inlier;  // Good matches (inlier feature matches)
    for (size_t i = 0; i < img_set.size(); i++)
    {
        for (size_t j = i + 1; j < img_set.size(); j++)
        {
            // Match features of two image pair (i, j) and find their inliers
            std::vector<cv::DMatch> match, inlier;
            fmatcher->match(img_descriptor[i], img_descriptor[j], match);
            std::vector<cv::Point2d> src, dst;
            for (auto itr = match.begin(); itr != match.end(); itr++)
            {
                src.push_back(img_keypoint[i][itr->queryIdx].pt);
                dst.push_back(img_keypoint[j][itr->trainIdx].pt);
            }
            cv::Mat inlier_mask;
            cv::findFundamentalMat(src, dst, inlier_mask, cv::RANSAC);
            for (int k = 0; k < inlier_mask.rows; k++)
                if (inlier_mask.at<uchar>(k)) inlier.push_back(match[k]);
            printf("3DV Tutorial: Image %zd - %zd are matched (%zd / %zd).\n", i, j, inlier.size(), match.size());

            // Determine whether the image pair is good or not
            if (inlier.size() < min_inlier_num) continue;
            printf("3DV Tutorial: Image %zd - %zd are selected.\n", i, j);
            match_pair.push_back(std::make_pair(uint(i), uint(j)));
            match_inlier.push_back(inlier);
            if (show_match)
            {
                cv::Mat match_image;
                cv::drawMatches(img_set[i], img_keypoint[i], img_set[j], img_keypoint[j], match, match_image, cv::Scalar::all(-1), cv::Scalar::all(-1), inlier_mask);
                cv::imshow("3DV Tutorial: Structure-from-Motion", match_image);
                cv::waitKey();
            }
        }
    }
    if (match_pair.size() < 1) return -1;

    // 1) Initialize cameras (rotation, translation, intrinsic parameters)
    std::vector<SFM::Vec9d> cameras(img_set.size(), SFM::Vec9d(0, 0, 0, 0, 0, 0, f_init, cx_init, cy_init));

    // 2) Initialize 3D points and build a visibility graph
    std::vector<cv::Point3d> Xs;
    std::vector<cv::Vec3b> Xs_rgb;
    SFM::VisibilityGraph xs_visited;
    for (size_t m = 0; m < match_pair.size(); m++)
    {
        for (size_t in = 0; in < match_inlier[m].size(); in++)
        {
            const uint &cam1_idx = match_pair[m].first, &cam2_idx = match_pair[m].second;
            const uint &x1_idx = match_inlier[m][in].queryIdx, &x2_idx = match_inlier[m][in].trainIdx;
            const uint key1 = SFM::genKey(cam1_idx, x1_idx), key2 = SFM::genKey(cam2_idx, x2_idx);
            auto visit1 = xs_visited.find(key1), visit2 = xs_visited.find(key2);
            if (visit1 != xs_visited.end() && visit2 != xs_visited.end())
            {
                // Remove previous observations if they are not consistent
                if (visit1->second != visit2->second)
                {
                    xs_visited.erase(visit1);
                    xs_visited.erase(visit2);
                }
                continue; // Skip if two observations are already visited
            }

            uint X_idx = 0;
            if (visit1 != xs_visited.end()) X_idx = visit1->second;
            else if (visit2 != xs_visited.end()) X_idx = visit2->second;
            else
            {
                // Add a new point if two observations are not visited
                X_idx = uint(Xs.size());
                Xs.push_back(cv::Point3d(0, 0, Z_init));
                Xs_rgb.push_back(img_set[cam1_idx].at<cv::Vec3b>(img_keypoint[cam1_idx][x1_idx].pt));
            }
            if (visit1 == xs_visited.end()) xs_visited[key1] = X_idx;
            if (visit2 == xs_visited.end()) xs_visited[key2] = X_idx;
        }
    }
    printf("3DV Tutorial: # of 3D points: %zd\n", Xs.size());

    // 3) Optimize camera pose and 3D points together (bundle adjustment)
    ceres::Problem ba;
    for (auto visit = xs_visited.begin(); visit != xs_visited.end(); visit++)
    {
        int cam_idx = SFM::getCamIdx(visit->first), x_idx = SFM::getObsIdx(visit->first);
        const cv::Point2d& x = img_keypoint[cam_idx][x_idx].pt;
        SFM::addCostFunc6DOF(ba, Xs[visit->second], x, cameras[cam_idx], ba_loss_width);
    }
    ceres::Solver::Options options;
    options.linear_solver_type = ceres::ITERATIVE_SCHUR;
    if (ba_num_iter > 0) options.max_num_iterations = ba_num_iter;
    options.num_threads = 8;
    options.minimizer_progress_to_stdout = true;
    ceres::Solver::Summary summary;
    ceres::Solve(options, &ba, &summary);
    std::cout << summary.FullReport() << std::endl;

    // Mark erroneous points to reject them
    std::vector<bool> is_noisy = maskNoisyPoints(Xs, img_keypoint, cameras, xs_visited, ba_loss_width);
    int num_noisy = std::accumulate(is_noisy.begin(), is_noisy.end(), 0);
    printf("3DV Tutorial: # of 3D points: %zd (Rejected: %d)\n", Xs.size(), num_noisy);
    for (size_t j = 0; j < cameras.size(); j++)
        printf("3DV Tutorial: Camera %zd's (f, cx, cy) = (%.3f, %.1f, %.1f)\n", j, cameras[j][6], cameras[j][7], cameras[j][8]);

    // Store the 3D points to an XYZ file
    FILE* fpts = fopen("sfm_global(point).xyz", "wt");
    if (fpts == NULL) return -1;
    for (size_t i = 0; i < Xs.size(); i++)
    {
        if (Xs[i].z > -Z_limit && Xs[i].z < Z_limit && !is_noisy[i])
            fprintf(fpts, "%f %f %f %d %d %d\n", Xs[i].x, Xs[i].y, Xs[i].z, Xs_rgb[i][2], Xs_rgb[i][1], Xs_rgb[i][0]); // Format: x, y, z, R, G, B
    }
    fclose(fpts);

    // Store the camera poses to an XYZ file 
    FILE* fcam = fopen("sfm_global(camera).xyz", "wt");
    if (fcam == NULL) return -1;
    for (size_t j = 0; j < cameras.size(); j++)
    {
        cv::Vec3d rvec(cameras[j][0], cameras[j][1], cameras[j][2]), t(cameras[j][3], cameras[j][4], cameras[j][5]);
        cv::Matx33d R;
        cv::Rodrigues(rvec, R);
        cv::Vec3d p = -R.t() * t;
        fprintf(fcam, "%f %f %f %f %f %f\n", p[0], p[1], p[2], R.t()(0, 2), R.t()(1, 2), R.t()(2, 2)); // Format: x, y, z, n_x, n_y, n_z
    }
    fclose(fcam);
    return 0;
}
